### ISSUE TITLE
FlowQueryList grabs an arbitrary thread's Looper

### ISSUE THREAD

**Issue by thirtythreeforty** [agrosner/DBFlow#1633]
The following line in 4.2.4 creates a Handler using the current thread's Looper:

https://github.com/agrosner/DBFlow/blob/0e06ea867c09db100a040314ca5127911afb16f9/dbflow/src/main/java/com/raizlabs/android/dbflow/list/FlowQueryList.java#L43

The equivalent line in `develop`:

https://github.com/agrosner/DBFlow/blob/c8306b8d15f6c34e94e0b20268f9ac24a37f99a1/lib/src/main/kotlin/com/dbflow5/query/list/FlowQueryList.kt#L205

There are two problems with this:
1. There is no guarantee that this initialization will run on the main thread
2. There is no guarantee that `Looper.myLooper()` will return an instance; it [might return null](https://stackoverflow.com/a/34322757/1830736).  (This is how I ran into this; see the stack trace below.)

I _believe_ this line is intended to make a static Handler that always runs things on the main thread.  If so, it should use `Looper.getMainLooper()`, which won't return null.

Stack trace leading me to this code:

```
Attempt to read from field 'android.os.MessageQueue android.os.Looper.mQueue' on a null object reference
       at android.os.Handler.(Handler.java:229)
       at android.os.Handler.(Handler.java:137)
       at com.raizlabs.android.dbflow.list.FlowQueryList.(Unknown Source:42)
       at com.raizlabs.android.dbflow.list.FlowQueryList$Builder.build(Unknown Source:734)
       at com.raizlabs.android.dbflow.sql.language.BaseModelQueriable.flowQueryList(Unknown Source:101)
    (snip)
```

**Comment by thirtythreeforty** [comment]
I think the scenario that triggers this issue is that the application starts up and the first thread to load the FlowQueryList class happens to be a worker thread that doesn't have a Looper.

**Comment by agrosner** [comment]
good catch! thank you. I think ill allow you to specify your own `Handler` too, but default to the main one.

**Comment by thirtythreeforty** [comment]
Great, thanks! Any chance you could backport to the 4.1 and/or 4.2 series? (Yay corporate code)

**Comment by whao189** [comment]
> ay corporate code

do you have resolved this question??

now, i used dbflow（4.2.4 release）, init database within  application's attachBaseContext....

then my another app（it's a demo）call this application by contentprovider,  throw exception  

“android.os.MessageQueue android.os.Looper.mQueue' on a null object reference” 。。。。。。

but , if i set init database within application's onCreate(context)....  the demo call this app by contentprovioder ,everythins is ok!! ....

@agrosner  @thirtythreeforty

**Comment by whao189** [comment]
i have a another question.   when i used dbflow  3.1.1 release（within application attachBaseContext funtion init database ）, modify the data and save it ..   then i export the database file by pragram（ through input stream operate ）on android 9.0 , open it by sqlite tools, but the modify data not change...... if i used dbflow 4.2.4 （of course ,within application onCreate funtion init database ）,modify app data ,....export the database file , open it by sqlite tools. thd modify data have bean changed ......

**Comment by thirtythreeforty** [comment]
> but , if i set init database within application's onCreate(context).... the demo call this app by contentprovioder ,everythins is ok!! ....

Yeah, the workaround for this is to make sure the class is loaded on the main thread.  The easiest way to do that would be to construct a throwaway instance during `Application#onCreate`.

**Comment by whao189** [comment]
> > but , if i set init database within application's onCreate(context).... the demo call this app by contentprovioder ,everythins is ok!! ....
> 
> Yeah, the workaround for this is to make sure the FlowQueryList class is loaded on the main thread. The easiest way to do that would be to construct a throwaway instance during `Application#onCreate`.

i have two apps, A and B .  
A is  a demo app, it call B through contentprovider.
B is my real app, it inculde a contentprovider, and provider's call method has querylistdata operate.

now , **B app is not startup , and app's process not exist.** then startup A app, clickbutton call B like this:
context.getContentResolver().call(B's URI, method,null);

the lifecycle of B, Application#attachBaseContext----ContentProvider#onCreate #Call() -----Application#onCreate(). 
question1 :  used dbflow 4.2.4 release . if init database during Applicaton#onCreate ,  my app crashed.....because ContentProvider#onCreate#Call lifecycle before of Applicaton#onCreate...  （B provider's call method has databasew query operate）and dbflow 4.2.4 not allowed init database during Application#attachBaseContext .     what shall i do ????  

question 2 : used dbflow 3.1.1 release, be allowed init database during Applicaton#attachBaseContext, but this version（3.1.1 release ）is not work well on Android9.0 , such as : modify data, save it to database ,then export the database file(through inputstream outputstream write file) . use SqliteTools open it , found the data not changed .....

i don't know , have you fixed them on DbFlow 5.0-beta ?????

**Comment by thirtythreeforty** [comment]
I find your formatting difficult to read, but as I understand it, the gist
is that the content provider thread starts before the main thread? If so,
then yes, you'll need a bug fix, which hasn't been posted here to my
knowledge. Alternatively, refactor your code not to use FlowQueryList,
which isn't as hard as it might sound.

Good luck whichever route you take!

**Comment by agrosner** [comment]
based on the original issue, it has been fixed in `develop` for 5.0.0-alpha2. any other questions or issues belong in a new ticket. thank you! as for backporting, it is not possible since Jitpack provides artifact ids based on the github organization. this was moved to my user after I had left the company for quite some time and I offered to continue maintaining it. So the artifact Id would need changing anyways.
